README.md
# libmachineid

Cross-platform C89 library to get a platform identifier, `libmachineid` will
attempt to gather a system identifier. If one cannot be determined a random
identifier is generated. The system identifier is hashed, and then optionally
truncated and converted to a UUIDv4 representation.

# Quick start

An example on Ubuntu:

```bash
sudo apt-get -y update
sudo apt-get install -y build-essential cmake git
git clone https://github.com/harporoeder/libmachineid.git
cd libmachineid
mkdir build
cmake ..
make
./test
> machine id: 4f6d8e5d-3837-473a-a707-8ec1f310e717
> error: MACHINEID_ERROR_NONE
```

## Error conditions

During utilization of `machineid_generate` an error may be returned of the
type `enum machineid_error`. Error conditions should be checked to ensure
that the result buffer will actually contain a valid result.

* `MACHINEID_ERROR_NONE` An identifier was successfully found.
* `MACHINEID_ERROR_RNG` An identifier was not found and the random number
generator also failed. No result is provided in this case.
* `MACHINEID_ERROR_NULL_OUTPUT_BUFFER` The library was utilized incorrectly,
and nowhere to store the result was indicated. No result is provided in this
case.
* `MACHINEID_ERROR_FALLBACK` An identifier was not found, but a fallback
was successfully generated
* `MACHINEID_HASH_FAILURE` The library utilized for hashing returned an error.
No result is provided in this case.

Error cases can be converted to a constant string with
`machineid_error_to_string`. This is likely useful for logging failures. The
library will never log of it's own volition.

## Flags

Bit flags of the type `enum machineid_flags` are provided to control the
desired behavior of `machineid_generate`.

* `MACHINEID_FLAG_DEFAULT` The default behavior of placing a raw digest in the
result buffer. This flag is only required when no other flags are used. When
using this mode the result buffer must be at least `MACHINEID_HASH_SIZE` in
size.
* `MACHINEID_FLAG_AS_UUID` Instead of returning a raw digest in the result 
buffer, truncate the digest and encode it as a UUIDv4. In this mode the
result buffer must be at least `MACHINEID_UUID_SIZE` in size.
* `MACHINEID_FLAG_NULL_TERMINATE` The character `'\0'` will be inserted at the
end of the result. When using this flag the result buffer must be either
`MACHINEID_HASH_SIZE + 1`, or `MACHINEID_UUID_SIZE + 1` as a minimum size
depending on the other flags used.

## Full example of library usage

```c
#include <stdio.h>

#include "machineid.h"

int main() {
    enum machineid_error err;
    unsigned char buffer[MACHINEID_UUID_SIZE + 1];

    err = machineid_generate(buffer, MACHINEID_FLAG_AS_UUID
        | MACHINEID_FLAG_NULL_TERMINATE);

    printf("error: %s\n", machineid_error_to_string(err));

    if (err != MACHINEID_ERROR_NONE && err != MACHINEID_ERROR_FALLBACK) {
        return 1;
    }
    
    printf("machine id: %s\n", buffer);

    return 0;
}
```

## Cryptography library integrations

For convenience `libmachineid` provides a vendored implementation of `SHA256`,
and platform specific random number generation capabilities. I recommend if
possible utilizing an external cryptography library. Support is available for
`sodium`, or `openssl`.

* To build with `sodium` use `cmake -D MACHINEID_USE_SODIUM=ON ..`
* To build with `openssl` use `cmake -D MACHINEID_USE_OPENSSL=ON ..`

When compiled with either of those options the vendored `SHA256` will not be
built, and random number generation will switch to using them.

# Platform support

The projected has been tested on Windows, MacOS, Linux, FreeBSD, and OpenBSD.
If a platform does not have first class support it will gracefully degrade with
reduced functionality.

If you need support for a specific platform that is not already handled open
an issue and it can likely be added contingent on my ability to test against it.

# Compiler support

The projected has been tested with Clang, MSVC, GCC, and TCC.

If you need support for additional compilers, and `libmachineid` does not
already build feel free to open an issue.

# Versioning

This project follows [semantic versioning](https://semver.org/). Additionally
the `main` branch is maintained in a usable state with development happening
on branches.

There exist 4 declarations inside `machineid.h` that can be used to
programmatically determine the version. There is a version string of the name
`MACHINEID_VERSION`, and 3 declarations for each specific component
`MACHINEID_VERSION_MAJOR`, `MACHINEID_VERSION_MINOR`,
and `MACHINEID_VERSION_PATCH`.

# Build system requirements

While [CMake](https://cmake.org/) is used in this repository it can be easily
avoided based on preference for an alternative build system.

The only notable complication is that you are required to define
`MACHINEID_USE_SODIUM`, or `MACHINEID_USE_OPENSSL` if you utilize either of
of those libraries with `libmachineid`.

# Sources of identifiers

## Linux

On Linux the file `/etc/machine-id` is used with `/var/lib/dbus/machine-id`
as an alternative.

## OpenBSD

On OpenBSD `HW_UUID` is queried. The file `/etc/machine-id` is used as an
alternative. DBUS is not installed by default and `/etc/machine-id` may not
be present.

## FreeBSD

On FreeBSD the file `/etc/hostid` is used.

## MacOS

On MacOS the registry key `kIOPlatformUUIDKey` is queried.

## Windows

On Windows the registry key `MachineGuid` is queried.

# Sources of entropy

When built with either `sodium` or `openssl` the random number generators
provided by those libraries are always used, otherwise a platform specific
generator is used.

On Windows `rand_s` is used. For OpenBSD or FreeBSD `arc4random_buf` is used.
On other platforms `rand` is used. When using `rand` you must ensure that you
seed with `srand` else fallback identifiers will be the same across
instances of your application.

# Considerations when utilizing Docker

It is common for Docker containers to have neither `/var/lib/dbus/machine-id`,
or `/etc/machine-id`. In this instance a fallback is generated. These files
can be made accessible to a container from the host system utilizing volumes:

```bash
docker run --rm -it \
    -v /etc/machine-id:/etc/machine-id:ro \
    ubuntu /bin/bash
```

This is not a universal property of Docker containers. Some may internally
run DBus, or otherwise provide a `machine-id`.

CHANGELOG.md
# Version 1.0.0 2021-01-03

Initial release.

test.c
/*
BSD 3-Clause License

Copyright (c) 2021, Harpo Roeder
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include "machineid.h"

#include <assert.h>
#include <stdio.h>
#include <string.h>

static void
test_error_string_encoding()
{
    assert(strcmp("MACHINEID_ERROR_NONE",
        machineid_error_to_string(MACHINEID_ERROR_NONE)) == 0);
    assert(strcmp("MACHINEID_ERROR_RNG",
        machineid_error_to_string(MACHINEID_ERROR_RNG)) == 0);
    assert(strcmp("MACHINEID_ERROR_NULL_OUTPUT_BUFFER",
        machineid_error_to_string(MACHINEID_ERROR_NULL_OUTPUT_BUFFER)) == 0);
    assert(strcmp("MACHINEID_ERROR_FALLBACK",
        machineid_error_to_string(MACHINEID_ERROR_FALLBACK)) == 0);
    assert(strcmp("MACHINEID_ERROR_HASH_FAILURE",
        machineid_error_to_string(MACHINEID_ERROR_HASH_FAILURE)) == 0);
    assert(machineid_error_to_string(52) == NULL);
}

static void
test_null_output_buffer()
{
    assert(MACHINEID_ERROR_NULL_OUTPUT_BUFFER == machineid_generate(
        NULL, MACHINEID_FLAG_DEFAULT));
}

static void
test_null_terminate_hash()
{
    unsigned char buffer[MACHINEID_HASH_SIZE + 1];

    buffer[MACHINEID_HASH_SIZE] = 52;

    machineid_generate(buffer, MACHINEID_FLAG_NULL_TERMINATE);

    assert(buffer[MACHINEID_HASH_SIZE] == '\0');
}

static void
test_null_terminate_uuid()
{
    unsigned char buffer[MACHINEID_UUID_SIZE + 1];

    buffer[MACHINEID_UUID_SIZE] = 52;

    machineid_generate(buffer, MACHINEID_FLAG_NULL_TERMINATE
        | MACHINEID_FLAG_AS_UUID);

    assert(buffer[MACHINEID_UUID_SIZE] == '\0');
}

static void
test_uuid_not_terminated()
{
    unsigned char buffer[MACHINEID_UUID_SIZE];

    memset(buffer, '\0', MACHINEID_UUID_SIZE);

    machineid_generate(buffer, MACHINEID_FLAG_AS_UUID);

    assert(buffer[MACHINEID_UUID_SIZE - 1] != '\0');
}

int
main()
{
    enum machineid_error err;
    unsigned char buffer[MACHINEID_UUID_SIZE + 1];

    test_error_string_encoding();
    test_null_output_buffer();
    test_null_terminate_hash();
    test_null_terminate_uuid();
    test_uuid_not_terminated();

    err = machineid_generate(buffer, MACHINEID_FLAG_AS_UUID
        | MACHINEID_FLAG_NULL_TERMINATE);

    printf("error: %s\n", machineid_error_to_string(err));

    if (err != MACHINEID_ERROR_NONE && err != MACHINEID_ERROR_FALLBACK) {
        return 1;
    }

    printf("machine id: %s\n", buffer);

    return 0;
}


sha256.c
/*********************************************************************
* Filename:   sha256.c
* Author:     Brad Conte (brad AT bradconte.com)
* Copyright:
* Disclaimer: This code is presented "as is" without any guarantees.
* Details:    Implementation of the SHA-256 hashing algorithm.
              SHA-256 is one of the three algorithms in the SHA2
              specification. The others, SHA-384 and SHA-512, are not
              offered in this implementation.
              Algorithm specification can be found here:
               * http://csrc.nist.gov/publications/fips/fips180-2/fips180-2withchangenotice.pdf
              This implementation uses little endian byte order.

This code is released into the public domain free of any restrictions.
The author requests acknowledgement if the code is used, but does not require
it. This code is provided free of any liability and without any quality claims
by the author.
*********************************************************************/

/*************************** HEADER FILES ***************************/
#include <stdlib.h>
#include <memory.h>
#include "sha256.h"

/****************************** MACROS ******************************/
#define ROTLEFT(a,b) (((a) << (b)) | ((a) >> (32-(b))))
#define ROTRIGHT(a,b) (((a) >> (b)) | ((a) << (32-(b))))

#define CH(x,y,z) (((x) & (y)) ^ (~(x) & (z)))
#define MAJ(x,y,z) (((x) & (y)) ^ ((x) & (z)) ^ ((y) & (z)))
#define EP0(x) (ROTRIGHT(x,2) ^ ROTRIGHT(x,13) ^ ROTRIGHT(x,22))
#define EP1(x) (ROTRIGHT(x,6) ^ ROTRIGHT(x,11) ^ ROTRIGHT(x,25))
#define SIG0(x) (ROTRIGHT(x,7) ^ ROTRIGHT(x,18) ^ ((x) >> 3))
#define SIG1(x) (ROTRIGHT(x,17) ^ ROTRIGHT(x,19) ^ ((x) >> 10))

/**************************** VARIABLES *****************************/
static const LIBSHA256_WORD k[64] = {
	0x428a2f98,0x71374491,0xb5c0fbcf,0xe9b5dba5,0x3956c25b,0x59f111f1,0x923f82a4,0xab1c5ed5,
	0xd807aa98,0x12835b01,0x243185be,0x550c7dc3,0x72be5d74,0x80deb1fe,0x9bdc06a7,0xc19bf174,
	0xe49b69c1,0xefbe4786,0x0fc19dc6,0x240ca1cc,0x2de92c6f,0x4a7484aa,0x5cb0a9dc,0x76f988da,
	0x983e5152,0xa831c66d,0xb00327c8,0xbf597fc7,0xc6e00bf3,0xd5a79147,0x06ca6351,0x14292967,
	0x27b70a85,0x2e1b2138,0x4d2c6dfc,0x53380d13,0x650a7354,0x766a0abb,0x81c2c92e,0x92722c85,
	0xa2bfe8a1,0xa81a664b,0xc24b8b70,0xc76c51a3,0xd192e819,0xd6990624,0xf40e3585,0x106aa070,
	0x19a4c116,0x1e376c08,0x2748774c,0x34b0bcb5,0x391c0cb3,0x4ed8aa4a,0x5b9cca4f,0x682e6ff3,
	0x748f82ee,0x78a5636f,0x84c87814,0x8cc70208,0x90befffa,0xa4506ceb,0xbef9a3f7,0xc67178f2
};

/*********************** FUNCTION DEFINITIONS ***********************/
void sha256_transform(SHA256_CTX *ctx, const LIBSHA256_BYTE data[])
{
	LIBSHA256_WORD a, b, c, d, e, f, g, h, i, j, t1, t2, m[64];

	for (i = 0, j = 0; i < 16; ++i, j += 4)
		m[i] = (data[j] << 24) | (data[j + 1] << 16) | (data[j + 2] << 8) | (data[j + 3]);
	for ( ; i < 64; ++i)
		m[i] = SIG1(m[i - 2]) + m[i - 7] + SIG0(m[i - 15]) + m[i - 16];

	a = ctx->state[0];
	b = ctx->state[1];
	c = ctx->state[2];
	d = ctx->state[3];
	e = ctx->state[4];
	f = ctx->state[5];
	g = ctx->state[6];
	h = ctx->state[7];

	for (i = 0; i < 64; ++i) {
		t1 = h + EP1(e) + CH(e,f,g) + k[i] + m[i];
		t2 = EP0(a) + MAJ(a,b,c);
		h = g;
		g = f;
		f = e;
		e = d + t1;
		d = c;
		c = b;
		b = a;
		a = t1 + t2;
	}

	ctx->state[0] += a;
	ctx->state[1] += b;
	ctx->state[2] += c;
	ctx->state[3] += d;
	ctx->state[4] += e;
	ctx->state[5] += f;
	ctx->state[6] += g;
	ctx->state[7] += h;
}

void sha256_init(SHA256_CTX *ctx)
{
	ctx->datalen = 0;
	ctx->bitlen = 0;
	ctx->state[0] = 0x6a09e667;
	ctx->state[1] = 0xbb67ae85;
	ctx->state[2] = 0x3c6ef372;
	ctx->state[3] = 0xa54ff53a;
	ctx->state[4] = 0x510e527f;
	ctx->state[5] = 0x9b05688c;
	ctx->state[6] = 0x1f83d9ab;
	ctx->state[7] = 0x5be0cd19;
}

void sha256_update(SHA256_CTX *ctx, const LIBSHA256_BYTE data[], size_t len)
{
	LIBSHA256_WORD i;

	for (i = 0; i < len; ++i) {
		ctx->data[ctx->datalen] = data[i];
		ctx->datalen++;
		if (ctx->datalen == 64) {
			sha256_transform(ctx, ctx->data);
			ctx->bitlen += 512;
			ctx->datalen = 0;
		}
	}
}

void sha256_final(SHA256_CTX *ctx, LIBSHA256_BYTE hash[])
{
	LIBSHA256_WORD i;

	i = ctx->datalen;

	/* Pad whatever data is left in the buffer. */
	if (ctx->datalen < 56) {
		ctx->data[i++] = 0x80;
		while (i < 56)
			ctx->data[i++] = 0x00;
	}
	else {
		ctx->data[i++] = 0x80;
		while (i < 64)
			ctx->data[i++] = 0x00;
		sha256_transform(ctx, ctx->data);
		memset(ctx->data, 0, 56);
	}

	/* Append to the padding the total message's length in bits and transform. */
	ctx->bitlen += (LIBSHA256_BYTE)(ctx->datalen * 8);
	ctx->data[63] = (LIBSHA256_BYTE)ctx->bitlen;
	ctx->data[62] = (LIBSHA256_BYTE)(ctx->bitlen >> 8);
	ctx->data[61] = (LIBSHA256_BYTE)(ctx->bitlen >> 16);
	ctx->data[60] = (LIBSHA256_BYTE)(ctx->bitlen >> 24);
	ctx->data[59] = (LIBSHA256_BYTE)(ctx->bitlen >> 32);
	ctx->data[58] = (LIBSHA256_BYTE)(ctx->bitlen >> 40);
	ctx->data[57] = (LIBSHA256_BYTE)(ctx->bitlen >> 48);
	ctx->data[56] = (LIBSHA256_BYTE)(ctx->bitlen >> 56);
	sha256_transform(ctx, ctx->data);

	/* Since this implementation uses little endian byte ordering and SHA uses big endian, */
	/* reverse all the bytes when copying the final state to the output hash. */
	for (i = 0; i < 4; ++i) {
		hash[i]      = (ctx->state[0] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 4]  = (ctx->state[1] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 8]  = (ctx->state[2] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 12] = (ctx->state[3] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 16] = (ctx->state[4] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 20] = (ctx->state[5] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 24] = (ctx->state[6] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 28] = (ctx->state[7] >> (24 - i * 8)) & 0x000000ff;
	}
}



sha256.h
/*********************************************************************
* Filename:   sha256.h
* Author:     Brad Conte (brad AT bradconte.com)
* Copyright:
* Disclaimer: This code is presented "as is" without any guarantees.
* Details:    Defines the API for the corresponding SHA1 implementation.

This code is released into the public domain free of any restrictions.
The author requests acknowledgement if the code is used, but does not require
it. This code is provided free of any liability and without any quality claims
by the author.
*********************************************************************/

#ifndef SHA256_H
#define SHA256_H

/*************************** HEADER FILES ***************************/
#include <stddef.h>
#include <stdint.h>

/****************************** MACROS ******************************/
#define SHA256_BLOCK_SIZE 32

/**************************** DATA TYPES ****************************/
typedef uint8_t LIBSHA256_BYTE;
typedef uint32_t LIBSHA256_WORD;

typedef struct {
	LIBSHA256_BYTE data[64];
	LIBSHA256_WORD datalen;
	uint64_t bitlen;
	LIBSHA256_WORD state[8];
} SHA256_CTX;

/*********************** FUNCTION DECLARATIONS **********************/
void sha256_init(SHA256_CTX *ctx);
void sha256_update(SHA256_CTX *ctx, const LIBSHA256_BYTE data[], size_t len);
void sha256_final(SHA256_CTX *ctx, LIBSHA256_BYTE hash[]);

#endif


machineid.h
/*
BSD 3-Clause License

Copyright (c) 2021, Harpo Roeder
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#ifndef MACHINEID_H
#define MACHINEID_H

#ifdef __cplusplus
extern "C" {
#endif

#define MACHINEID_VERSION "1.0.0"
#define MACHINEID_VERSION_MAJOR 1
#define MACHINEID_VERSION_MINOR 0
#define MACHINEID_VERSION_PATCH 0

#define MACHINEID_HASH_SIZE 32
#define MACHINEID_UUID_SIZE 36

enum machineid_flags {
    MACHINEID_FLAG_DEFAULT        = 0,
    MACHINEID_FLAG_AS_UUID        = 1,
    MACHINEID_FLAG_NULL_TERMINATE = 2
};

enum machineid_error {
    MACHINEID_ERROR_NONE               = 0,
    MACHINEID_ERROR_RNG                = 1,
    MACHINEID_ERROR_NULL_OUTPUT_BUFFER = 2,
    MACHINEID_ERROR_FALLBACK           = 3,
    MACHINEID_ERROR_HASH_FAILURE       = 4
};

const char *machineid_error_to_string(const enum machineid_error err);

enum machineid_error machineid_generate(unsigned char *const outputBuffer,
    const enum machineid_flags flags);

#ifdef __cplusplus
}
#endif

#endif


machineid.c
/*
BSD 3-Clause License

Copyright (c) 2021, Harpo Roeder
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <stdio.h>
#include <string.h>
#include <stddef.h>

#ifdef _WIN32
#define _CRT_RAND_S
#include <windows.h>
#endif

#include <stdlib.h>

#ifdef __APPLE__
#include <CoreFoundation/CoreFoundation.h>
#include <IOKit/IOKitLib.h>
#endif

#ifdef MACHINEID_USE_SODIUM
#include <sodium.h>
#include <openssl/rand.h>
#elif MACHINEID_USE_OPENSSL
#include <openssl/sha.h>
#else
#include "sha256.h"
#endif

#ifdef __OpenBSD__
#include <sys/param.h>
#include <sys/sysctl.h>
#endif

#include "machineid.h"

static void machineid_bin_to_hex(unsigned char *const outputBuffer,
    const unsigned char *const inputBuffer, const size_t inputBufferSize);

static void machineid_bin_to_uuid(unsigned char *const outputBuffer,
    const unsigned char *const inputBuffer);

static size_t machineid_raw(unsigned char *const outputBuffer,
    const size_t outputBufferSize);

static char machineid_random_bytes(unsigned char *const outputBuffer,
    const size_t count);

static char machineid_sha256(unsigned char *const outputBuffer,
    const unsigned char *const inputBuffer, const size_t inputBufferSize);

const char *const HEX_ALPHABET = "0123456789abcdef";

#ifndef MIN
#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))
#endif

static char
machineid_random_bytes(unsigned char *const outputBuffer, const size_t count)
{
#ifdef MACHINEID_USE_SODIUM
    randombytes_buf((void *const)outputBuffer, count);

    return 0;
#elif MACHINEID_USE_OPENSSL
    return RAND_bytes(outputBuffer, (int)count) != 0;
#elif defined(__OpenBSD__) || defined(__FreeBSD__)
    arc4random_buf((void *const)outputBuffer, count);

    return 0;
#elif _WIN32
    size_t i;
    errno_t err;

    for (i = 0; i < count; i++) {
        unsigned int number;

        err = rand_s(&number);

        if (err != 0) {
            return 1;
        }

        outputBuffer[i] = number;
    }

    return 0;
#else
    size_t i;

    for (i = 0; i < count; i++) {
        outputBuffer[i] = rand();
    }

    return 0;
#endif
}

static char
machineid_sha256(unsigned char *const outputBuffer,
    unsigned const char *const inputBuffer,
    const size_t inputBufferSize)
{
#ifdef MACHINEID_USE_SODIUM
    return crypto_hash_sha256(outputBuffer, inputBuffer, inputBufferSize);
#elif MACHINEID_USE_OPENSSL
    SHA256(inputBuffer, inputBufferSize, outputBuffer);

    return 0;
#else
    SHA256_CTX context;

    sha256_init(&context);
    sha256_update(&context, (const LIBSHA256_BYTE *const)inputBuffer,
        inputBufferSize);
    sha256_final(&context, (LIBSHA256_BYTE *const)outputBuffer);

    return 0;
#endif
}

enum machineid_error
machineid_generate(unsigned char *const outputBuffer,
    const enum machineid_flags flags)
{
    char fallback;
    unsigned char rawBuffer[256], hashBuffer[MACHINEID_HASH_SIZE];
    size_t rawSize;
    int status;

    fallback = 0;

    if (outputBuffer == NULL) {
        return MACHINEID_ERROR_NULL_OUTPUT_BUFFER;
    }

    rawSize = machineid_raw(rawBuffer, sizeof(rawBuffer));

    if (rawSize == 0) {
        if (machineid_random_bytes(rawBuffer, 16)) {
            return MACHINEID_ERROR_RNG;
        } else {
            fallback = 1;
        }
    }

    status = machineid_sha256(hashBuffer, rawBuffer, rawSize);

    if (status != 0) {
        return MACHINEID_ERROR_HASH_FAILURE;
    }

    if (flags & MACHINEID_FLAG_AS_UUID) {
        machineid_bin_to_uuid(outputBuffer, hashBuffer);
    } else {
        memcpy(outputBuffer, hashBuffer, sizeof(hashBuffer));
    }

    if (flags & MACHINEID_FLAG_NULL_TERMINATE) {
        if (flags & MACHINEID_FLAG_AS_UUID) {
            outputBuffer[MACHINEID_UUID_SIZE] = '\0';
        } else {
            outputBuffer[MACHINEID_HASH_SIZE] = '\0';
        }
    }

    if (fallback == 1) {
        return MACHINEID_ERROR_FALLBACK;
    } else {
        return MACHINEID_ERROR_NONE;
    }
}

#if defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__)
static size_t
posix_read_file(const char *const path, unsigned char *const outputBuffer,
    const size_t outputBufferSize)
{
    FILE *handle;
    long fileSize;
    int status;
    size_t expectedResultSize, resultSize;

    handle = fopen(path, "r");

    if (handle == NULL) {
        return 0;
    }

    status = fseek(handle, 0L, SEEK_END);

    if (status != 0) {
        goto err;
    }

    fileSize = ftell(handle);

    if (fileSize == -1L) {
        goto err;
    }

    status = fseek(handle, 0L, SEEK_SET);

    if (status != 0) {
        goto err;
    }

    expectedResultSize = MIN((size_t)fileSize, outputBufferSize);

    resultSize = fread(outputBuffer, sizeof(unsigned char),
        expectedResultSize, handle);
    
    if (resultSize != expectedResultSize) {
        goto err;
    }

    fclose(handle);

    return resultSize;

  err:
    fclose(handle);

    return 0;
}
#endif

size_t
machineid_raw(unsigned char *const outputBuffer, const size_t outputBufferSize)
{
#ifdef __linux__
    size_t resultSize;

    resultSize = posix_read_file("/etc/machine-id", outputBuffer,
        outputBufferSize);

    if (resultSize != 0) {
        return resultSize;
    }

    return posix_read_file("/var/lib/dbus/machine-id", outputBuffer,
        outputBufferSize);
#elif __FreeBSD__
    return posix_read_file("/etc/hostid", outputBuffer,
        outputBufferSize);
#elif __OpenBSD__
    int mib[2], status;
    size_t len;

    mib[0] = CTL_HW;
    mib[1] = HW_UUID;

    status = sysctl(mib, 2, NULL, &len, NULL, 0);

    if (status == -1) {
        goto fallback;
    }

    if (len >= outputBufferSize) {
        goto fallback;
    }

    status = sysctl(mib, 2, outputBuffer, &len, NULL, 0);

    if (status != -1) {
        goto fallback;
    }

    return len;

  fallback:
    return posix_read_file("/etc/machine-id", outputBuffer,
        outputBufferSize);
#elif __APPLE__
    io_registry_entry_t registryEntry;
    CFStringRef identifier;
    Boolean status;

    registryEntry = IORegistryEntryFromPath(kIOMasterPortDefault,
        "IOService:/");

    if (registryEntry == MACH_PORT_NULL) {
        return 0;
    }

    identifier = (CFStringRef)IORegistryEntryCreateCFProperty(registryEntry,
        CFSTR(kIOPlatformUUIDKey), kCFAllocatorDefault, 0);

    IOObjectRelease(registryEntry);

    if (identifier == NULL) {
        return 0;
    }

    status = CFStringGetCString(identifier, (char *const)outputBuffer,
        (CFIndex)outputBufferSize, kCFStringEncodingASCII);

    if (status == false) {
         CFRelease(identifier);

         return 0;
    }

    CFRelease(identifier);

    return strlen((const char *const)outputBuffer) + 1;
#elif _WIN32
    LSTATUS status;
    HKEY key;
    DWORD lpType, lpcbData;

    status = RegOpenKeyExA(HKEY_LOCAL_MACHINE,
        "SOFTWARE\\Microsoft\\Cryptography", 0,
        KEY_READ | KEY_WOW64_64KEY, &key);

    if (status != ERROR_SUCCESS) {
        return 0;
    }

    lpType = REG_SZ;
    lpcbData = (DWORD)outputBufferSize;

    status = RegQueryValueExA(key, "MachineGuid", NULL, &lpType, outputBuffer,
        &lpcbData);

    if (status != ERROR_SUCCESS) {
        RegCloseKey(key);

        return 0;
    }

    return (size_t)lpcbData;
#else
    return 0;
#endif
}

static void
machineid_bin_to_hex(unsigned char *const outputBuffer,
    const unsigned char *const inputBuffer, const size_t inputBufferSize)
{
    size_t i;
    unsigned char *outputIter;
    const unsigned char *inputIter;

    outputIter = outputBuffer;
    inputIter = inputBuffer;

    for (i = 0; i < inputBufferSize; i += 1) {
        *outputIter++ = HEX_ALPHABET[(*inputIter>>4)&0xF];
        *outputIter++ = HEX_ALPHABET[(*inputIter++)&0xF];
    }
}

static void
machineid_bin_to_uuid(unsigned char *const outputBuffer,
    const unsigned char *const inputBuffer)
{
    unsigned char *outputIter;
    const unsigned char *inputIter;
    unsigned int i;

    outputIter = outputBuffer;
    inputIter = inputBuffer;

    machineid_bin_to_hex(outputIter, inputIter, 4);

    outputIter += 8;
    inputIter += 4;
    *outputIter = '-';
    outputIter++;

    for (i = 0; i < 3; i++) {
        machineid_bin_to_hex(outputIter, inputIter, 2);

        outputIter += 4;
        inputIter += 2;
        *outputIter = '-';
        outputIter++;
    }

    machineid_bin_to_hex(outputIter, inputIter, 6);
}

const char *
machineid_error_to_string(const enum machineid_error err)
{
    switch (err) {
        case MACHINEID_ERROR_NONE:
            return "MACHINEID_ERROR_NONE";
            break;

        case MACHINEID_ERROR_RNG:
            return "MACHINEID_ERROR_RNG";
            break;

        case MACHINEID_ERROR_NULL_OUTPUT_BUFFER:
            return "MACHINEID_ERROR_NULL_OUTPUT_BUFFER";
            break;

        case MACHINEID_ERROR_FALLBACK:
            return "MACHINEID_ERROR_FALLBACK";
            break;

        case MACHINEID_ERROR_HASH_FAILURE:
            return "MACHINEID_ERROR_HASH_FAILURE";
            break;
    }

    return NULL;
}
